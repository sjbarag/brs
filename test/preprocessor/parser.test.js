const brs = require("brs");
const preprocessor = brs.preprocessor;
const { Lexeme } = brs.lexer;
const { BrsBoolean } = brs.types;

describe("preprocessor parser", () => {
    let parser;

    beforeEach(() => {
        parser = new preprocessor.Parser();
    });

    it("parses chunks of brightscript", () => {
        let { chunks, errors } = parser.parse([
            { kind: Lexeme.Identifier, text: "someFunction", line: 1, isReserved: false },
            { kind: Lexeme.LeftParen, text: "(", line: 1, isReserved: false },
            { kind: Lexeme.RightParen, text: ")", line: 1, isReserved: false },
            { kind: Lexeme.Newline, text: "\n", line: 1, isReserved: false },
            { kind: Lexeme.Eof, text: "\0", line: 2, isReserved: true }
        ]);

        expect(errors).toEqual([]);
        expect(chunks).toBeDefined();
        expect(chunks).not.toBeNull();
        expect(chunks).toMatchSnapshot();
    });

    it("parses #const", () => {
        let { chunks, errors } = parser.parse([
            { kind: Lexeme.HashConst, text: "#const", line: 1, isReserved: false },
            { kind: Lexeme.Identifier, text: "foo", line: 1, isReserved: false },
            { kind: Lexeme.Equal, text: "=", line: 1, isReserved: false },
            { kind: Lexeme.True, text: "true", literal: BrsBoolean.True, line: 1, isReserved: true },
            { kind: Lexeme.Newline, text: "\n", line: 1, isReserved: false },
            { kind: Lexeme.Eof, text: "\0", line: 1, isReserved: false }
        ]);

        expect(errors).toEqual([]);
        expect(chunks).toBeDefined();
        expect(chunks).not.toBeNull();
        expect(chunks).toMatchSnapshot();
    });

    it("parses #error", () => {
        let { chunks, errors } = parser.parse([
            { kind: Lexeme.HashError, text: "#error", line: 1, isReserved: false },
            { kind: Lexeme.HashErrorMessage, text: "I'm an error message!", line: 1, isReserved: false },
            { kind: Lexeme.Eof, text: "\0", line: 1, isReserved: true }
        ]);

        expect(errors).toEqual([]);
        expect(chunks).toBeDefined();
        expect(chunks).not.toBeNull();
        expect(chunks).toMatchSnapshot();
    });

    describe("conditionals", () => {
        test("#if only", () => {
            let { chunks, errors } = parser.parse([
                { kind: Lexeme.HashIf, text: "#if", isReserved: false, location: { start: { line: 1, column: 1 }, end: { line: 1, column: 3 } } },
                { kind: Lexeme.Identifier, text: "foo", line: 1, isReserved: false },
                { kind: Lexeme.Newline, text: "\n", line: 1, isReserved: false },
                { kind: Lexeme.Identifier, text: "fooIsTrue", line: 2, isReserved: false },
                { kind: Lexeme.LeftParen, text: "(", line: 2, isReserved: false },
                { kind: Lexeme.RightParen, text: ")", line: 2, isReserved: false },
                { kind: Lexeme.Newline, text: "\n", line: 2, isReserved: false },
                { kind: Lexeme.HashEndIf, text: "#endif", line: 3, isReserved: false },
                { kind: Lexeme.Eof, text: "\0", line: 3, isReserved: false }
            ]);

            expect(errors).toEqual([]);
            expect(chunks).toBeDefined();
            expect(chunks).not.toBeNull();
            expect(chunks).toMatchSnapshot();
        });

        test("#if and #else", () => {
            let { chunks, errors } = parser.parse([
                { kind: Lexeme.HashIf, text: "#if", isReserved: false, location: { start: { line: 1, column: 1 }, end: { line: 1, column: 3 } } },
                { kind: Lexeme.Identifier, text: "foo", line: 1, isReserved: false },
                { kind: Lexeme.Newline, text: "\n", line: 1, isReserved: false },

                { kind: Lexeme.Identifier, text: "fooIsTrue", line: 2, isReserved: false },
                { kind: Lexeme.LeftParen, text: "(", line: 2, isReserved: false },
                { kind: Lexeme.RightParen, text: ")", line: 2, isReserved: false },
                { kind: Lexeme.Newline, text: "\n", line: 2, isReserved: false },

                { kind: Lexeme.HashElse, text: "#else", line: 3, isReserved: false },
                { kind: Lexeme.Newline, text: "\n", line: 3, isReserved: false },

                { kind: Lexeme.Identifier, text: "fooIsFalse", line: 4, isReserved: false },
                { kind: Lexeme.LeftParen, text: "(", line: 4, isReserved: false },
                { kind: Lexeme.RightParen, text: ")", line: 4, isReserved: false },
                { kind: Lexeme.Newline, text: "\n", line: 4, isReserved: false },

                { kind: Lexeme.HashEndIf, text: "#endif", line: 5, isReserved: false },
                { kind: Lexeme.Eof, text: "\0", line: 5, isReserved: false }
            ]);

            expect(errors).toEqual([]);
            expect(chunks).toBeDefined();
            expect(chunks).not.toBeNull();
            expect(chunks).toMatchSnapshot();
        });

        test("#if #else if and #else", () => {
            let { chunks, errors } = parser.parse([
                { kind: Lexeme.HashIf, text: "#if", isReserved: false, location: { start: { line: 1, column: 1 }, end: { line: 1, column: 3 } } },
                { kind: Lexeme.Identifier, text: "foo", line: 1, isReserved: false },
                { kind: Lexeme.Newline, text: "\n", line: 1, isReserved: false },

                { kind: Lexeme.Identifier, text: "fooIsTrue", line: 2, isReserved: false },
                { kind: Lexeme.LeftParen, text: "(", line: 2, isReserved: false },
                { kind: Lexeme.RightParen, text: ")", line: 2, isReserved: false },
                { kind: Lexeme.Newline, text: "\n", line: 2, isReserved: false },

                { kind: Lexeme.HashElseIf, text: "#elseif", line: 3, isReserved: false },
                { kind: Lexeme.Identifier, text: "bar", line: 3, isReserved: false },
                { kind: Lexeme.Newline, text: "\n", line: 3, isReserved: false },

                { kind: Lexeme.Identifier, text: "bar", line: 4, isReserved: false },
                { kind: Lexeme.LeftParen, text: "(", line: 4, isReserved: false },
                { kind: Lexeme.RightParen, text: ")", line: 4, isReserved: false },
                { kind: Lexeme.Newline, text: "\n", line: 4, isReserved: false },

                { kind: Lexeme.HashElse, text: "#else", line: 5, isReserved: false },
                { kind: Lexeme.Newline, text: "\n", line: 5, isReserved: false },

                { kind: Lexeme.Identifier, text: "neither", line: 6, isReserved: false },
                { kind: Lexeme.LeftParen, text: "(", line: 6, isReserved: false },
                { kind: Lexeme.RightParen, text: ")", line: 6, isReserved: false },
                { kind: Lexeme.Newline, text: "\n", line: 6, isReserved: false },

                { kind: Lexeme.HashEndIf, text: "#endif", line: 7, isReserved: false },
                { kind: Lexeme.Eof, text: "\0", line: 7, isReserved: false }
            ]);

            expect(errors).toEqual([]);
            expect(chunks).toBeDefined();
            expect(chunks).not.toBeNull();
            expect(chunks).toMatchSnapshot();
        });
    });
});
