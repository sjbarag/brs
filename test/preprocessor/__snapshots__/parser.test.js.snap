// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`preprocessor parser conditionals #if #else if and #else 1`] = `
Array [
  If {
    "condition": Object {
      "isReserved": false,
      "kind": 21,
      "line": 1,
      "text": "foo",
    },
    "elseChunk": BrightScript {
      "tokens": Array [
        Object {
          "isReserved": false,
          "kind": 21,
          "line": 6,
          "text": "neither",
        },
        Object {
          "isReserved": false,
          "kind": 0,
          "line": 6,
          "text": "(",
        },
        Object {
          "isReserved": false,
          "kind": 1,
          "line": 6,
          "text": ")",
        },
        Object {
          "isReserved": false,
          "kind": 84,
          "line": 6,
          "text": "
",
        },
      ],
    },
    "elseIfs": Array [
      Object {
        "condition": Object {
          "isReserved": false,
          "kind": 21,
          "line": 3,
          "text": "bar",
        },
        "thenChunk": BrightScript {
          "tokens": Array [
            Object {
              "isReserved": false,
              "kind": 84,
              "line": 3,
              "text": "
",
            },
            Object {
              "isReserved": false,
              "kind": 21,
              "line": 4,
              "text": "bar",
            },
            Object {
              "isReserved": false,
              "kind": 0,
              "line": 4,
              "text": "(",
            },
            Object {
              "isReserved": false,
              "kind": 1,
              "line": 4,
              "text": ")",
            },
            Object {
              "isReserved": false,
              "kind": 84,
              "line": 4,
              "text": "
",
            },
          ],
        },
      },
    ],
    "thenChunk": BrightScript {
      "tokens": Array [
        Object {
          "isReserved": false,
          "kind": 21,
          "line": 2,
          "text": "fooIsTrue",
        },
        Object {
          "isReserved": false,
          "kind": 0,
          "line": 2,
          "text": "(",
        },
        Object {
          "isReserved": false,
          "kind": 1,
          "line": 2,
          "text": ")",
        },
        Object {
          "isReserved": false,
          "kind": 84,
          "line": 2,
          "text": "
",
        },
      ],
    },
  },
]
`;

exports[`preprocessor parser conditionals #if and #else 1`] = `
Array [
  If {
    "condition": Object {
      "isReserved": false,
      "kind": 21,
      "line": 1,
      "text": "foo",
    },
    "elseChunk": BrightScript {
      "tokens": Array [
        Object {
          "isReserved": false,
          "kind": 21,
          "line": 4,
          "text": "fooIsFalse",
        },
        Object {
          "isReserved": false,
          "kind": 0,
          "line": 4,
          "text": "(",
        },
        Object {
          "isReserved": false,
          "kind": 1,
          "line": 4,
          "text": ")",
        },
        Object {
          "isReserved": false,
          "kind": 84,
          "line": 4,
          "text": "
",
        },
      ],
    },
    "elseIfs": Array [],
    "thenChunk": BrightScript {
      "tokens": Array [
        Object {
          "isReserved": false,
          "kind": 21,
          "line": 2,
          "text": "fooIsTrue",
        },
        Object {
          "isReserved": false,
          "kind": 0,
          "line": 2,
          "text": "(",
        },
        Object {
          "isReserved": false,
          "kind": 1,
          "line": 2,
          "text": ")",
        },
        Object {
          "isReserved": false,
          "kind": 84,
          "line": 2,
          "text": "
",
        },
      ],
    },
  },
]
`;

exports[`preprocessor parser conditionals #if only 1`] = `
Array [
  If {
    "condition": Object {
      "isReserved": false,
      "kind": 21,
      "line": 1,
      "text": "foo",
    },
    "elseChunk": undefined,
    "elseIfs": Array [],
    "thenChunk": BrightScript {
      "tokens": Array [
        Object {
          "isReserved": false,
          "kind": 21,
          "line": 2,
          "text": "fooIsTrue",
        },
        Object {
          "isReserved": false,
          "kind": 0,
          "line": 2,
          "text": "(",
        },
        Object {
          "isReserved": false,
          "kind": 1,
          "line": 2,
          "text": ")",
        },
        Object {
          "isReserved": false,
          "kind": 84,
          "line": 2,
          "text": "
",
        },
      ],
    },
  },
]
`;

exports[`preprocessor parser parses #const 1`] = `
Array [
  Declaration {
    "name": Object {
      "isReserved": false,
      "kind": 21,
      "line": 1,
      "text": "foo",
    },
    "value": Object {
      "isReserved": true,
      "kind": 81,
      "line": 1,
      "literal": BrsBoolean {
        "kind": 1,
        "value": true,
      },
      "text": "true",
    },
  },
]
`;

exports[`preprocessor parser parses #error 1`] = `
Array [
  Error {
    "hashError": Object {
      "isReserved": false,
      "kind": 36,
      "line": 1,
      "text": "#error",
    },
    "message": "I'm an error message!",
  },
]
`;

exports[`preprocessor parser parses chunks of brightscript 1`] = `
Array [
  BrightScript {
    "tokens": Array [
      Object {
        "isReserved": false,
        "kind": 21,
        "line": 1,
        "text": "someFunction",
      },
      Object {
        "isReserved": false,
        "kind": 0,
        "line": 1,
        "text": "(",
      },
      Object {
        "isReserved": false,
        "kind": 1,
        "line": 1,
        "text": ")",
      },
      Object {
        "isReserved": false,
        "kind": 84,
        "line": 1,
        "text": "
",
      },
    ],
  },
]
`;
